### DISCUSSION TITLE
Plans on dropping Windows x86 (32-bit) binaries

### DISCUSSION THREAD

**Discussion by vszakats** [curl/curl-for-win#68] [Ideas]
Opening this thread to discuss and schedule dropping 32-bit official 
Windows curl binaries, leaving x64 and ARM64.

This would make development leaner, builds faster, scripts simpler.

Non-Windows targets already support the latter two only, plus 
RISCV64 as an experiment.

Most hardware sold in the last decade+ is 64-bit compatible (including
Atom). Windows Vista / Server 2008 (the minimum our builds require)
do support 64-bit. Windows 10 was the last version supporting 32-bit.
MS stopped offering 32-bit Windows 10 to OEMs in 2020. Windows 10
is no longer supported after October 14, 2025. Windows 11 is 64-bit only.

Cygwin [dropped](https://cygwin.com/pipermail/cygwin-announce/2022-November/010777.html) x86 support in November 2022. Most Linux distros did so earlier.

Git for Windows dropped 32-bit for its full packages:
https://github.com/git-for-windows/git/releases/tag/v2.49.0.windows.1
Also from Chocolatey: https://github.com/chocolatey-community/chocolatey-packages/commit/4bad18d0a1319c8a0490ced370a532859801350b

Windows 11 ARM64 edition can run x64 binaries (according to my
tests). This is important for Chocolatey, because it doesn't support
native ARM64 binaries yet: https://github.com/chocolatey/choco/issues/1803 
https://github.com/chocolatey/choco/issues/2172. I haven't checked how it behaves in this 
specific scenario. There are already Chocolatey projects that started
dropping 32-bit binaries.

winget and scoop both support native ARM64.

x86/x64 stats for Firefox envs:
https://data.firefox.com/dashboard/hardware#operating-system-metric-overview-2

Windows x86 vs x64 vs ARM64 in the annual curl survey
- [2025](https://curl.se/docs/survey/2025-1.1/#windows-versions) 32-bit: 7% [PDF](https://daniel.haxx.se/media/survey-2025.pdf)
- [2024](https://daniel.haxx.se/media/curl-user-survey-2024-analysis.pdf) 32-bit: 11.2%

Other projects:
https://www.msys2.org/news/#2023-12-13-starting-to-drop-some-32-bit-packages

My proposed, tentative, schedule for this is mid-2025.

edit to clarify: this thread is solely about offering official 32-bit
Windows curl/libcurl _binaries_ as part of the project.
Building curl for Windows 32-bit remains possible and there are
no plans to change it.

**Comment by jay** [comment]
This is one of those things I don't think we're going to really know the impact until you do it. I know I've seen some corporate monitors used to mandate 32-bit so they could analyze or hook into programs and there wasn't a 64-bit counterpart. That was many years ago though. I would ask on the lists if anyone is using the official 32-bit curl for Windows instead of 64-bit, and why.

edit: save stats for curl for Windows 32-bit and 64-bit downloads might help

**Comment by vszakats** [comment]
Download stats would be excellent. My understanding is that the CDN doesn't count inidividual URLs/downloads, only aggregate traffic, but if there is way to pull this info, it'd help.

Also asking the mailing list, indeed. Would you do it?

Chocolatey has download stats, but it's not broken down by CPU:
https://community.chocolatey.org/packages/curl#versionhistory

(with the caveat that downloads don't necessary match the runtime
env, 32-bit curl works just fine on an x64 machine. But a useful 
indicator nevertheless.)

**Reply by bagder** [reply]
> Download stats would be excellent. My understanding is that the CDN doesn't count individual URLs/downloads, only aggregate traffic, but if there is way to pull this info, it'd help.

Correct, we don't have any logging from the CDN so all data we have is accumulated for the site and it does not separate between different downloads.

We could perhaps consider doing the Windows download specifically use a separate (non-CDNed) "bounce host" for a period of time just to gather this kind of statistics. (ie change the download links to `https://download.curl.se/$file` that then simply redirects the agent to the final and real target URL)

**Reply by vszakats** [reply]
It would catch direct downloads, but would not those via package managers or other automated ways, unless those get updated to the new endpoints (which in turn can cause confusion when the experiment ends). Could still be useful data.

Or, we catch the URL with a web server rule (?) and redirect it to the measured endpoint. This has some risk of breaking downloaders not expecting redirection, but would catch every download.

Another angle to consider is that we also distribute libcurl, for use as a dependency for applications. Meaning those apps will also need to be upgraded to 64-bit, along with their other dependencies, which may or may not support 64-bit. I know of at least one such anecdotal case, where a dependency is stuck with the commercial developer shipping 32-bit binaries only. Source code could also have vanished with a 32-bit binary remaining. As for how many people are using these libcurl binaries is of anyone's best guess. IMO not many, but in this case counting downloads isn't indicative, because a developer downloads once, then deploys the dependent app many times.

Staggered deprecation may help, first announce, then drop 32-bit builds and downloads – while keeping 32-bit build curl-for-win logic. Wait for feedback. Then remove the logic at a later point.

**Reply by jay** [reply]
> Also asking the mailing list, indeed. Would you do it?

Sent to curl-users: [If you're using an official curl for Windows 32-bit build instead of 64-bit please tell us why](https://curl.se/mail/archive-2024-03/0007.html)

**Reply by vszakats** [reply]
Thank you Jay!

**Comment by vszakats** [comment]
Number of curl issues/PRs with mentions:
- 64-bit: 140x (https://github.com/curl/curl/issues?q=%22x86_64-w64-mingw32%22)
- 32-bit: 45x (https://github.com/curl/curl/issues?q=%22i686-w64-mingw32%22)

**Comment by vszakats** [comment]
32-bit `curl.exe`: Can run on any hardware and any Windows that is Vista/2008 or newer.
64-bit `curl.exe`: Requires 64-bit hardware running 64-bit Windows Vista/2008 or newer.

32-bit libcurl static lib and DLL: Can only link to a 32-bit application.
64-bit libcurl static lib and DLL: Can only link to a 64-bit application.

**Comment by killerbees19** [comment]
> Building curl for Windows 32-bit remains possible and there are no plans to change it.

@vszakats Do you mean building via `curl-for-win` repo remains possible after 2025? Or only by hand without this great repo?

Because I'm using self-built libcurl.dll for x86 and I'd be screwed without this repo. Sadly it's impossible to update the program to x64, it's closed source abandonware with reverse-engineered patches, even for initial cURL support.

**Reply by vszakats** [reply]
@killerbees19 Thanks for your feeback.

Yes, this is the unfixable scenario briefly mentioned above.
Also a reminder that these scripts are used locally too.

It is an option to keep the x86 build logic, and only disable x86
builds/downloads by default.

This path will continue to require x86-specific maintenance, 
and e.g. extending/testing new build options for x86. We can 
lean more on the community for these maybe.

**Reply by killerbees19** [reply]
Yeah, I get the point that just dropping the official downloads without removing the actual code/maintenance burden isn't really helping long-term. (And I really appreciate your time-consuming work for this project/repo!)

Maybe an "unofficial" state for x86 (self-builds) could be feasible? I mean something like "if it breaks someone [else] needs to fix it" and "not explicitly tested for each cURL release". Because (cross) building cURL (and SSL libs) by hand without this repo is hard. Been there, tried it, failed and switched to the already existing solution. 😅

Sure, at some point I could fork this repo and try to keep it up-to-date for further cURL release at x86. But I'm pretty sure that'd be a big fail after some weeks/months without support. I've not enough knowledge and experience in this area. Last but not least there's no customer or employer who could pay for this stuff, I need it only for nonprofit projects used by ~1000 people around Europe. Most likely as last resort I'd just stick to an old cURL version and hope for the best...

Anyways, I respect and understand your upcoming decisions around x86, whatever they'll be. Just a big "I'M STILL USING IT" feedback from my side, even if I'm not using the original downloads. 🙂

**Reply by vszakats** [reply]
Thanks @killerbees19!

I understand, and can relate to your situation, also knowing such cases first/second-hand. I did discuss ways to "integrate" a 64-bit binary with parts stuck at 32-bit, meaning decoupling them into separate binaries, make them communicate/RPC, or execute each other as bare external binaries. These may be doable, but the effort is non-trivial and not always an option, depending. Maybe something to think about long term.

Agreed, that an unofficial status can be viable. It would help some with daily maintenance, binary sizes, CI times. And, effectively mean a little bit more involvement from the community if anything breaks due to no first-hand testing; then see how that goes.

FWIW curl-for-win already supports a bunch of features that aren't part of the default or official binaries.

**Comment by vszakats** [comment]
I propose dropping 32-bit official Windows binaries on 2025-Aug-20 (curl release day).

We keep 32-bit as a non-default, deprecated, option. The `-x86` build flag will be necessary to enable it. There will be no active testing of this on our infrastructure. We will be relying on community testing and help in case of fallouts.

**Comment by vszakats** [comment]
@bagder Pinging you to possibly include a survey question of which Windows CPU platform(s) one is using: 64-bit Intel, 32-bit Intel, 64-bit ARM. Re: https://curl.se/mail/lib-2024-05/0008.html

**Reply by bagder** [reply]
I'm adding this question:

## If you use curl on Windows, which architecture?

- 64 bit Intel
- 32 bit Intel
- 64 bit ARM
- I don't know

**Reply by bagder** [reply]
With multiple answers allowed (even if "I don't know" is weird to combine with the others)

**Reply by killerbees19** [reply]
@vszakats @bagder Maybe `x86` would be a better wording? Otherwise AMD users could mistakenly select the last option, even if it's the same architecture.

* x86 (32 bit)
* x86-64 (64 bit)
* ARM (64 bit)
* I don't know

**Reply by vszakats** [reply]
Possible, or what about both? As for `x86-64` I think the Windowsy terminology is `x64`. So something like?:

* Intel 64-bit (x64)
* Intel 32-bit (x86)
* ARM 64-bit
* I don't know

**Reply by bagder** [reply]
yeah, I think @vszakats' version here looks the best

**Comment by vszakats** [comment]
Thanks to @bagder, and everyone filling the curl 2024 survey, here is the CPU platform breakdown:
![a-012](https://github.com/curl/curl-for-win/assets/1446897/c620e648-a195-43be-8fea-263df412bc9a)

Blog post: https://daniel.haxx.se/blog/2024/06/17/curl-user-survey-2024-analysis/
Full report, direct link: https://daniel.haxx.se/media/curl-user-survey-2024-analysis.pdf

With a surprise ARM64 share exceeding Intel 32-bit.
Mozilla stats for comparison: https://data.firefox.com/dashboard/hardware#operating-system-metric-overview-2

**Reply by jay** [reply]
I always think it will be lower. 10% on 32-bit still represents a lot of users.

**Reply by vszakats** [reply]
It matches up with the Mozilla number: 10.997% (as of 2024-06-10)

This after a recent 3% drop when Firefox apparently started detecting ARM64.

Another potential (perhaps the best) data point could be this: https://github.com/curl/curl-www/issues/346

**Comment by vszakats** [comment]
32-bit share update on 2025-07-03:
Mozilla: 9.010%
2025 curl survey: 7%
